storage/series.go

```diff
@@ -233,6 +233,79 @@ func (s *dbSeries) ReadEncoded(
 	return results, nil
 }
 
+func (s *dbSeries) FetchBlocks(ctx context.Context, starts []time.Time) []FetchBlockResult {
+	var res []FetchBlockResult
+
+	s.RLock()
+
+	for _, start := range starts {
+		if b, exists := s.blocks.GetBlockAt(start); exists {
+			stream, err := b.Stream(ctx)
+			if err != nil {
+				detailedErr := fmt.Errorf("unable to retrieve block stream for series %s time %v: %v", s.seriesID, start, err)
+				res = append(res, newFetchBlockResult(start, nil, detailedErr))
+			} else if stream != nil {
+				res = append(res, newFetchBlockResult(start, []xio.SegmentReader{stream}, nil))
+			}
+		}
+	}
+
+	if !s.buffer.Empty() {
+		bufferResults := s.buffer.FetchBlocks(ctx, starts)
+		res = append(res, bufferResults...)
+	}
+
+	s.RUnlock()
+
+	sortFetchBlockResultByTimeAscending(res)
+
+	return res
+}
+
+func (s *dbSeries) FetchBlocksMetadata(ctx context.Context, includeSizes bool) (block.DatabaseBlocksMetadata, error) {
+	var (
+		res      []block.DatabaseBlockMetadata
```

Comment:
For now this is only called at the beginning of peer streaming, which only happens during bootstrapping, so I'd rather let them GC'ed than keeping them around in a pool. I guess in the future if we need background repairs we might need to call this function semi-frequently, yeah can add the TODO.